### PR TITLE
Debug mode

### LINKED ISSUE #457: Debug Mode
One of the most appealing point of Chainer is easiness of debugging. But verbose assertions can cause much overhead. 

This PR proposes debug mode during forward and backward operations.
Possible feature includes
- [x] Detection of `inf` and `nan`
- [x] More assertion of the condition (e.g. #302)
- [x] Check of inconsistency in `grad` and `data` (cf. #278)
- [ ] Visualize errors by computational graph (cf. #252)
* ...

### PR DESCRIPTION
**Pull request by unnonouno**
I added debug mode. It raises an error when nan is detected.

### CODE REVIEW

File: chainer/function.py [hunk 1]
```diff
@@ -98,6 +100,9 @@ def __call__(self, *inputs):
 
         """
         in_data = tuple([x.data for x in inputs])
+        if chainer.DEBUG:
+            self._stack = traceback.extract_stack()
```
Comment (beam2d) [inline chainer/function.py]: Is this trace back not used? How about showing it on errors in backprop?

File: chainer/__init__.py [hunk 2]
```diff
@@ -29,4 +29,6 @@
 OFF = flag.OFF
 AUTO = flag.AUTO
 
+DEBUG = False
```
Comment (beam2d) [inline chainer/__init__.py]: Is it intended to share the DEBUG flag across multiple threads?

File: tests/chainer_tests/test_function.py [hunk 3]
```diff
@@ -316,4 +316,50 @@ def test_forward_invalid(self):
...
+    @attr.gpu
+    def test_debug_forward_gpu(self):
+        self.f.forward_gpu = mock.MagicMock(
+            return_value=(cuda.to_gpu(self.nan),))
+        self.check_debug_forward(cuda.to_gpu(self.one))
+
+    def check_debug_backward(self, x_data):
+        x = chainer.Variable(x_data)
+        with self.assertRaises(RuntimeError):
+            y = self.f(x)
```
Comment (beam2d) [inline tests/chainer_tests/test_function.py]: Move the forward computation to outside of the assertRaises block.

File: chainer/function.py (hunk 1 above)
Comment (unnonouno) [inline chainer/function.py]: I'll fix it after #934 is merged

**Comment by unnonouno** [comment]
We decided to use a method to set/get the debug mode instead of using a global variable.

File: chainer/__init__.py [hunk 4]
```diff
@@ -29,4 +29,26 @@
 OFF = flag.OFF
 AUTO = flag.AUTO
 
+_debug = False
+
+
+def get_debug():
```
Comment (beam2d) [inline chainer/__init__.py]: is_debug is better

**Comment by beam2d** [comment]
Add is_debug / set_debug to the reference.

**Comment by beam2d** [comment]
LGTM!

**Pull request merged**